python/cudf/cudf/core/udf/groupby_function.py

```diff
@@ -0,0 +1,851 @@
+    ):  # fe_type is fully instantiated group type
+        members = [
+            ("group_data", types.CPointer(fe_type.group_scalar_type)),
+            ("size", types.int64),
+            ("index", types.CPointer(fe_type.index_type)),
+        ]
+        models.StructModel.__init__(self, dmm, fe_type, members)
+
+
+my_max_int32 = cuda.declare_device(
+    "BlockMax_int32", "types.int32(types.CPointer(types.int32),types.int64)"
+)
+
+my_max_int64 = cuda.declare_device(
+    "BlockMax_int64", "types.int64(types.CPointer(types.int64),types.int64)"
+)
+
+my_max_float64 = cuda.declare_device(
+    "BlockMax_float64",
+    "types.float64(types.CPointer(types.float64),types.int64)",
+)
+
+my_min_int32 = cuda.declare_device(
+    "BlockMin_int32", "types.int32(types.CPointer(types.int32),types.int64)"
+)
+
+my_min_int64 = cuda.declare_device(
+    "BlockMin_int64", "types.int64(types.CPointer(types.int64),types.int64)"
+)
+
+my_min_float64 = cuda.declare_device(
+    "BlockMin_float64",
+    "types.float64(types.CPointer(types.float64),types.int64)",
+)
+
+my_sum_int64 = cuda.declare_device(
+    "BlockSum_int64", "types.int64(types.CPointer(types.int64),types.int64)"
+)
+
+my_sum_float64 = cuda.declare_device(
+    "BlockSum_float64",
+    "types.float64(types.CPointer(types.float64),types.int64)",
+)
+
+my_mean_int64 = cuda.declare_device(
+    "BlockMean_int64",
+    "types.float64(types.CPointer(types.int64),types.int64)",
+)
+
+my_mean_float64 = cuda.declare_device(
+    "BlockMean_float64",
+    "types.float64(types.CPointer(types.float64),types.int64)",
+)
+
+my_std_int64 = cuda.declare_device(
+    "BlockStd_int64",
+    "types.float64(types.CPointer(types.int64),types.int64)",
+)
+
+my_std_float64 = cuda.declare_device(
+    "BlockStd_float64",
+    "types.float64(types.CPointer(types.float64),types.int64)",
+)
+
+my_var_int64 = cuda.declare_device(
+    "BlockVar_int64",
+    "types.float64(types.CPointer(types.int64),types.int64)",
+)
+
+my_var_float64 = cuda.declare_device(
+    "BlockVar_float64",
+    "types.float64(types.CPointer(types.float64),types.int64)",
+)
+
+my_idxmax_int64 = cuda.declare_device(
+    "BlockIdxMax_int64",
+    "types.int64(types.CPointer(types.int64),"
+    "types.CPointer(types.int64),types.int64)",
+)
+
+my_idxmax_float64 = cuda.declare_device(
+    "BlockIdxMax_float64",
+    "types.int64(types.CPointer(types.float64),"
+    "types.CPointer(types.int64),types.int64)",
+)
+
+my_idxmin_int64 = cuda.declare_device(
+    "BlockIdxMin_int64",
+    "types.int64(types.CPointer(types.int64),"
+    "types.CPointer(types.int64),types.int64)",
+)
+
+my_idxmin_float64 = cuda.declare_device(
+    "BlockIdxMin_float64",
+    "types.int64(types.CPointer(types.float64),"
+    "types.CPointer(types.int64),types.int64)",
+)
+
+# Path to the source containing the foreign function
+basedir = os.path.dirname(os.path.realpath(__file__))
+dev_func_ptx = os.path.join(basedir, "function.ptx")
+
+
+def call_my_max_int32(data, size):
```

Comment:
What does this layer of indirection actually do? It should be equivalent to pass `my_max_int32` anywhere that you can pass `call_my_max_int32`.

Reply:
This wraps the object returned by `cuda.declare_device` into a standard python function which is what `compile_internal`  expects down the line.